### PR TITLE
Supervision with a BehaviorInterceptor

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala [hunk 1]
```diff
@@ -737,9 +820,8 @@ class SupervisionSpec extends ScalaTestWithActorTestKit(
 
       actor ! "give me stacktrace"
       val stacktrace = probe.expectMessageType[Vector[StackTraceElement]]
-      // supervisor receive is used for every supervision instance, only wrapped in one supervisor for RuntimeException
-      // and then the IllegalArgument one is kept since it has a different throwable
-      stacktrace.count(_.toString.startsWith("akka.actor.typed.internal.Supervisor.receive")) should ===(2)
+      stacktrace.foreach(println)
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala]: Remove? I vaguely remember adding a custom failure message to a scalatest assert, but I can't remember how that was done.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+      case r: Resume ⇒
+        Behaviors.intercept[T, T](new ResumeSupervisor(r))(initialBehavior)
+      case r: Restart ⇒
+        Behaviors.intercept[T, T](new RestartSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Stop ⇒
+        Behaviors.intercept[T, T](new StopSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Backoff ⇒
+        Behaviors.intercept[AnyRef, T](new BackoffSupervisor(initialBehavior, r))(initialBehavior).asInstanceOf[Behavior[T]]
+    }
+  }
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: So much cleaner than the old one, nice!

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+abstract class AbstractSupervisor[O, I, Thr <: Throwable](ss: SupervisorStrategy)(implicit ev: ClassTag[Thr]) extends BehaviorInterceptor[O, I] {
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: `@InternalApi private[akka]`, for all these classes

`strategy` would be nicer than `ss`

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    }
+  }
+
+  protected def handleExceptionOnStart(ctx: ActorContext[O]): Catcher[Behavior[I]]
+  protected def handleSignalException(ctx: ActorContext[O], target: BehaviorInterceptor.SignalTarget[I]): Catcher[Behavior[I]]
+  protected def handleReceiveException(ctx: ActorContext[O], target: BehaviorInterceptor.ReceiveTarget[I]): Catcher[Behavior[I]]
+}
+
+/**
+ * For cases where O == I for BehaviorInterceptor.
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: ?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+ */
+abstract class SimpleSupervisor[T, Thr <: Throwable: ClassTag](ss: SupervisorStrategy) extends AbstractSupervisor[T, T, Thr](ss) {
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Is this expected to be public API? If not, fold it into AbstractSupervisor instead?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    case NonFatal(t: Thr) ⇒
+      log(ctx, t)
+      Behaviors.same
+  }
+}
+
+class RestartSupervisor[T, Thr <: Throwable](initial: Behavior[T], strategy: Restart)(implicit ev: ClassTag[Thr]) extends SimpleSupervisor[T, Thr](strategy) {
+
+  var restarts = 0
+  var deadline: OptionVal[Deadline] = OptionVal.None
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: private?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+  private def restart(ctx: ActorContext[_], t: Throwable) = {
+    val timeLeft = deadlineHasTimeLeft
+    val newDeadline = if (deadline.isDefined && timeLeft) deadline else OptionVal.Some(Deadline.now + strategy.withinTimeRange)
+    restarts = if (timeLeft) restarts + 1 else 1
+    deadline = newDeadline
+  }
+
+  private def handleException(ctx: ActorContext[T], signalRestart: () ⇒ Unit): Catcher[Behavior[T]] = {
+    case NonFatal(t: Thr) ⇒
+      println(s"ex: $t. $restarts $deadlineHasTimeLeft $strategy")
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: printlns are for line printers

**Review by johanandren — COMMENTED** [review]
Some small things, looks great in general.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 6 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Yes i need to do a round of privatisation

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 7 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: dammit my secret is out that I use printlns for debugging

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala [hunk 8]
```diff
@@ -315,6 +316,18 @@ class SupervisionSpec extends ScalaTestWithActorTestKit(
       }
     }
 
+    "stop when strategy is stop - exception in setup" in {
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala]: isn't this true also without stop strategy, with restart?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 2 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: good name change also

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        Behaviors.intercept[T, T](new RestartSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Stop ⇒
+        Behaviors.intercept[T, T](new StopSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Backoff ⇒
+        Behaviors.intercept[AnyRef, T](new BackoffSupervisor(initialBehavior, r))(initialBehavior).asInstanceOf[Behavior[T]]
+    }
+  }
+}
+
+private abstract class AbstractSupervisor[O, I, Thr <: Throwable](strategy: SupervisorStrategy)(implicit ev: ClassTag[Thr]) extends BehaviorInterceptor[O, I] {
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: how private is this for Java? perhaps just add the internal annotations

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    } catch handleSignalException(ctx, target)
+  }
+
+  def log(ctx: ActorContext[_], t: Throwable): Unit = {
+    if (strategy.loggingEnabled) {
+      ctx.asScala.log.error(t, "Supervisor [{}] saw failure: {}", this, t.getMessage)
+    }
+  }
+
+  protected def handleExceptionOnStart(ctx: ActorContext[O]): Catcher[Behavior[I]]
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: I love `Catcher`, used too seldom

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+  def aroundSignal(ctx: ActorContext[O], signal: Signal, target: SignalTarget[I]): Behavior[I] = {
+    try {
+      target(ctx, signal)
+    } catch handleSignalException(ctx, target)
+  }
+
+  def log(ctx: ActorContext[_], t: Throwable): Unit = {
+    if (strategy.loggingEnabled) {
+      ctx.asScala.log.error(t, "Supervisor [{}] saw failure: {}", this, t.getMessage)
+    }
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: superfluous curlies

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+ */
+private abstract class SimpleSupervisor[T, Thr <: Throwable: ClassTag](ss: SupervisorStrategy) extends AbstractSupervisor[T, T, Thr](ss) {
+
+  override def aroundReceive(ctx: ActorContext[T], msg: T, target: BehaviorInterceptor.ReceiveTarget[T]): Behavior[T] = {
+    try {
+      target(ctx, msg)
+    } catch handleReceiveException(ctx, target)
+  }
+
+  protected def handleException(ctx: ActorContext[T]): Catcher[Behavior[T]] = {
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: should this be abstract?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+/**
+ * For cases where O == I for BehaviorInterceptor.
+ */
+private abstract class SimpleSupervisor[T, Thr <: Throwable: ClassTag](ss: SupervisorStrategy) extends AbstractSupervisor[T, T, Thr](ss) {
+
+  override def aroundReceive(ctx: ActorContext[T], msg: T, target: BehaviorInterceptor.ReceiveTarget[T]): Behavior[T] = {
+    try {
+      target(ctx, msg)
+    } catch handleReceiveException(ctx, target)
+  }
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: why is this here and not in AbstractSupervisor?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+  override def aroundStart(ctx: ActorContext[T], target: BehaviorInterceptor.PreStartTarget[T]): Behavior[T] = {
+    try {
+      target.start(ctx)
+    } catch {
+      case NonFatal(t: Thr) ⇒
+        // if unlimited restarts then don't restart if starting fails as it would likely be an infinite restart loop
+        log(ctx, t)
+        if (strategy.unlimitedRestarts() || ((restarts + 1) >= strategy.maxNrOfRetries && deadlineHasTimeLeft)) {
+          throw t
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: is this what takes care of stopping when exception is thrown from "constructor" (top level `setup`)?
does that result in double logging at the place where it is caught?

**Review by patriknw — COMMENTED** [review]
very nice, some smallish things...

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala (hunk 8 above)
Comment (chbatey) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala]: Yes, only difference is the log. Maybe that shouldn't be logged as it'll the the cause of the ActorIntializationException

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 9 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Can't be too private!

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 12 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: It could be. `RestartSupervisor` doesn't use it as it needs the targets for dealing with exceptions so it would need to provide an unused impl

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 13 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: The types don't line up there as it has two type params I for the inner and O for the outter. This simple impl only works when I == O. Otherwise the message type is O and the target is I.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 14 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Yes in that case we'd get a log from the supervisor and then an ActorIntiializationException logged with the exception as the cause. Shall i remove the one in the supervisor?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 4 above)
Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: When the Outer and Inner types are the same (true for all the of the supervisors apart from backoff as it schedules other messages to be sent to its self) we can provide a defaiult impl for aroundReceive

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 14 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: um, double logging is never good

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: but add a comment

Comment (chbatey) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: :+1:

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 4 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Ahaa, that explains the need for it.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
Fixes #25602
